robot_calibration/src/capture_features.cpp

```diff
@@ -0,0 +1,304 @@
+// Author: Jack Center
+
+#include <fstream>
+#include <ctime>
+
+#include <ros/ros.h>
+#include <rosbag/bag.h>
+#include <rosbag/view.h>
+#include <rosbag/query.h>
+
+#include <std_msgs/String.h>
+#include <robot_calibration_msgs/CalibrationData.h>
+#include <robot_calibration_msgs/CaptureConfig.h>
+
+#include <robot_calibration/capture/chain_manager.h>
+#include <robot_calibration/capture/feature_finder_loader.h>
+
+#include <camera_calibration_parsers/parse.h>
+#include <robot_calibration/ceres/optimizer.h>
+#include <robot_calibration/camera_info.h>
+#include <robot_calibration/load_bag.h>
+
+#include <boost/foreach.hpp>  // for rosbag iterator
+
+#include <robot_calibration/capture_features.h>
+
+int main(int argc, char** argv)
+{
+  ros::init(argc, argv,"robot_calibration");
+  ros::NodeHandle nh("~");
+
+  if (!check_parameters(nh))
+    return -1;
+
+  // Set up the calibration
+// TODO: make timeout a define
+  robot_calibration::ChainManager chain_manager(nh, 0.1);     // Manages kinematic chains
+
+  robot_calibration::FeatureFinderMap finders;               // Holds the available feature finders  
+  if (!get_feature_finders(nh, finders)) 
+    return -1;    
+
+  std::string feature;
+  nh.getParam("feature_finder", feature);                    // The feature to search for during caputre
+
+  std::string bag_filename;                                   // Absolute path to location to save the bagfile
+  nh.getParam("bag_filename", bag_filename);  
+  rosbag::Bag bag(bag_filename, rosbag::bagmode::Write);
+
+  // Get the robot_description and republish it
+// TODO: describe why this is needed.
+  std_msgs::String description_msg;
+  nh.getParam("/robot_description", description_msg.data);
+  ros::Publisher urdf_pub = nh.advertise<std_msgs::String>("/robot_description", 1, true);  // latched
+  urdf_pub.publish(description_msg);
+  bag.write("/robot_description", ros::Time::now(), description_msg);
+
+  bool capture_successful;
+
+  // Auto capture mode
+  if (nh.param<bool>("auto_calibration_mode", false))
+  {
+    ROS_INFO("Using automatic calibration mode...");
+    capture_successful = run_automatic_capture(nh, &chain_manager, finders, feature, bag);
+  }
+
+  // Manual capture mode
+  else
+  {
+    ROS_INFO("Using manual calibration mode...");
+    capture_successful = run_manual_capture(nh, &chain_manager, finders, feature, bag);
+  }
+
+  bag.close();
+  ROS_INFO("Done capturing samples");
+
+  return 0;
+}
+
+bool check_parameters(ros::NodeHandle& nh)
```

Comment:
oh, this is the check_parameters.  In this case you should put this in a private namespace above the main to make it clear that this is the check_parameters function for this file.